EmbeddedSurfacesParallelSynchronization.cpp

```diff
@@ -0,0 +1,435 @@
+    {
+      if( edgeGhostsToSend[a] == parentIndex[ni] )
+      {
+        // a node is sent if the edge on which it was created had to be sent.
+        newNodesToSend.emplace_back( ni );
+      }
+    } );
+  }
+
+  newElemsToSendData.resize( elemManager.numRegions() );
+  newElemsToSend.resize( elemManager.numRegions() );
+  //
+  if( newObjects.newElements.size() > 0 )  // this map may be completely empty on some ranks.
+  {
+    elemManager.forElementSubRegionsComplete< EmbeddedSurfaceSubRegion >(
+      [&]( localIndex const er, localIndex const esr, ElementRegionBase &, EmbeddedSurfaceSubRegion & subRegion )
+    {
+
+      FixedToManyElementRelation const & surfaceElementsToCells = subRegion.getToCellRelation();
+
+      for( localIndex const & k : newObjects.newElements.at( {er, esr} ) )
+      {
+        localIndex const elemRegionIndex  = surfaceElementsToCells.m_toElementRegion[k][0];
+        localIndex const elemSubRegionIndex = surfaceElementsToCells.m_toElementSubRegion[k][0];
+        localIndex const elemIndex = surfaceElementsToCells.m_toElementIndex[k][0];
+
+        CellElementRegion const & elemRegion = elemManager.getRegion< CellElementRegion >( elemRegionIndex );
+        CellBlock const & elemSubRegion = elemRegion.getSubRegion< CellElementSubRegion >( elemSubRegionIndex );
+
+        arrayView1d< localIndex const > const & elemGhostsToSend =
+          elemSubRegion.getNeighborData( neighborRank ).ghostsToSend();
+
+        forAll< serialPolicy >( elemGhostsToSend.size(), [=, &newSurfaceGhostsToSend] ( localIndex const a )
+        {
+          if( elemIndex == elemGhostsToSend[a] )
+          {
+            newSurfaceGhostsToSend[ {er, esr} ].emplace_back( k );
+            return;
+          }
+        } );
+      }
+    } );
+  }
+
+  for( localIndex er=0; er<elemManager.numRegions(); ++er )
+  {
+    ElementRegionBase & elemRegion = elemManager.getRegion( er );
+    newElemsToSendData[er].resize( elemRegion.numSubRegions() );
+    newElemsToSend[er].resize( elemRegion.numSubRegions() );
+    if( newSurfaceGhostsToSend.size() > 0 )  // this map may be completely empty on some ranks.
+    {
+      elemRegion.forElementSubRegionsIndex< EmbeddedSurfaceSubRegion >( [&]( localIndex const esr,
+                                                                             EmbeddedSurfaceSubRegion & subRegion )
+      {
+        localIndex_array & surfaceElemGhostsToSend = subRegion.getNeighborData( neighborRank ).ghostsToSend();
+
+        std::vector< localIndex >  newGhostsElements = newSurfaceGhostsToSend.at( {er, esr} );
+        forAll< serialPolicy >( newGhostsElements.size(), [=, &newElemsToSendData, &surfaceElemGhostsToSend] (localIndex const k)
+        {
+          newElemsToSendData[er][esr].emplace_back( newGhostsElements[k] );
+          surfaceElemGhostsToSend.emplace_back( newGhostsElements[k] );
+        } ) ;
+        newElemsToSend[er][esr] = newElemsToSendData[er][esr];
+      } );
```

Comment:
@corbett5 is this kernel correct? I was not sure what to do with that std::vector.

Reply:
Unless you've moved it previously you should move `surfaceElemGhostsToSend` to the CPU. You can't capture it by view since you need to call `emplace_back`.

As for the std::vector you should change it to `std::vector< localIndex > const &` so that you're not making a copy of it. But then you actually copy it again when you capture it by value in the lambda. So you should capture it by reference. At this point though all three variables used in the lambda are captured by reference so it's not getting you anything (not moving any views).